finagle/h2/src/main/scala/com/twitter/finagle/buoyant/h2/ResponseClassifiers.scala

```diff
@@ -68,41 +68,87 @@ object ResponseClassifiers {
   }
 
   /**
+   * Create a [[ResponseClassifier]] with the given name for its `toString`.
+   */
+  def named(name: String)(underlying: ResponseClassifier): ResponseClassifier =
+    new ResponseClassifier {
+      def isDefinedAt(reqRep: H2ReqRep): Boolean = underlying.isDefinedAt(reqRep)
+      def apply(reqRep: H2ReqRep): ResponseClass = underlying(reqRep)
+      override def toString: String = name
+    }
+
+  /**
    * Classifies 5XX responses as failures. If the method is idempotent
    * (as described by RFC2616), it is classified as retryable.
    */
   val RetryableIdempotentFailures: ResponseClassifier =
-    ResponseClassifier.named("RetryableIdempotentFailures") {
-      case ReqRep(Requests.Idempotent(), RetryableResult()) => ResponseClass.RetryableFailure
+    named("RetryableIdempotentFailures") {
+      case H2ReqRep(Requests.Idempotent(), RetryableResult()) => ResponseClass.RetryableFailure
     }
 
   /**
    * Classifies 5XX responses as failures. If the method is a read
    * operation, it is classified as retryable.
    */
   val RetryableReadFailures: ResponseClassifier =
-    ResponseClassifier.named("RetryableReadFailures") {
-      case ReqRep(Requests.ReadOnly(), RetryableResult()) => ResponseClass.RetryableFailure
+    named("RetryableReadFailures") {
+      case H2ReqRep(Requests.ReadOnly(), RetryableResult()) => ResponseClass.RetryableFailure
     }
 
   /**
    * Classifies 5XX responses and all exceptions as non-retryable
    * failures.
    */
   val NonRetryableServerFailures: ResponseClassifier =
-    ResponseClassifier.named(s"NonRetryableServerFailures") {
-      case ReqRep(_, Throw(_) | Return(Responses.Failure())) => ResponseClass.NonRetryableFailure
+    named(s"NonRetryableServerFailures") {
+      case H2ReqRep(_, Throw(_) | Return((Responses.Failure(), _))) => ResponseClass.NonRetryableFailure
```

Comment:
I think we want to classify `(_, Return((_, Throw(_))))` as a failure too.  eg when we get a 200 OK but the response stream fails.